### PR TITLE
feat: merge operators for commutative LSM operations

### LINKED ISSUE #15: feat: merge operators for commutative LSM operations
## Upstream Reference

- fjall-rs/lsm-tree#104

## Why CoordiNode Needs This

Posting-list append/remove operations on graph edges are commutative. Without merge operators, edge mutations require read-modify-write cycles, which cause OCC abort storms on super-nodes (>100K edges).

CoordiNode architecture (`arch/core/principles.md`) explicitly calls out read-modify-write as an anti-pattern and mandates `storage.merge()` operators instead of GET -> modify -> PUT.

## Scope

1. **Core merge operator** — `MergeOperator` trait, `ValueType::MergeOperand`, merge-aware reads/compaction
2. **Range tombstone interaction** — merge resolution skips RT-suppressed entries (treated as deletion boundaries) in both point reads (`resolve_merge_get`) and range scans (`MvccStream`)
3. **Indirection (BlobTree) safety** — all merge paths detect `ValueType::Indirection` and skip merge resolution (blob pointer bytes are not user data)

## Disk Format Impact

This feature is **not** just an in-memory API addition.

Merge operators introduce a new persisted `ValueType::MergeOperand` in the LSM value encoding. Older `V3` readers do not understand that value kind and do not implement merge-resolution semantics during reads or compaction.

That means merge operators must participate in the fork's shared **disk format V4** boundary together with range tombstones:

- `V3` remains the legacy format for older databases
- `V4` is required for databases written with persisted merge operands and/or range tombstones
- older `V3` binaries must reject these databases instead of opening them and producing incorrect results

## Priority

**P0** — blocks efficient graph edge mutations

## Status

Implemented in fork branch / PR #28.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Add `MergeOperator` trait and persisted `ValueType::MergeOperand` support for commutative LSM operations
- Merge-aware point reads, range scans, and compaction
- Range tombstone awareness in merge resolution (point reads, multi_get, and range scans)
- Enables efficient graph edge mutations without read-modify-write cycles

## Technical Details

**New persisted type**: `ValueType::MergeOperand = 3` stores partial updates (operands) that are lazily combined via a user-provided `MergeOperator` trait.

**Write path**: `tree.merge(key, operand, seqno)` appends a `MergeOperand` entry to the memtable.

**Read path**: `tree.get()` and `tree.multi_get()` collect all entries for the key across memtable, sealed memtables, and disk tables. The disk scan collects all visible entries for the key across all runs (handling L0 overlap correctly by sorting in descending seqno order), then applies the merge function in chronological order (ascending seqno). Tombstones and RT-suppressed entries stop the search and produce a merge with no base value. When no merge operator is configured, both `get()` and `multi_get()` check RT suppression before returning raw operand bytes.

**Range tombstone interaction**: `resolve_merge_get` (point reads) checks each entry against `is_suppressed_by_range_tombstones` before including it in the merge. `MvccStream` (range scans) accepts optional RT state with per-source visibility cutoffs via `with_range_tombstones()` and skips RT-suppressed entries during merge collection (treated as deletion boundaries). If the newest `MergeOperand` itself is RT-suppressed, merge resolution is skipped entirely. The post-MvccStream `RangeTombstoneFilter` remains for non-merge entries.

**Compaction**: `CompactionStream` merges operands below the GC watermark via the merge operator. When a base value or tombstone boundary is found in the same stream, the result is a `Value` (complete merge). When no boundary is found (partial merge — base may be in a lower level), the result stays as `MergeOperand` to avoid shadowing the real base. Operands above the watermark are preserved for MVCC. Compaction filters that replace a `MergeOperand` with a `Value` preserve the `MergeOperand` type to prevent base shadowing (Indirection replacement is allowed through to avoid storing blob pointers under the MergeOperand type). Seqno zeroing is skipped for preserved `MergeOperand` entries to avoid creating duplicate internal keys.

**Indirection (BlobTree) safety**: All merge resolution paths (MvccStream forward/reverse, CompactionStream, resolve_merge_get) detect `ValueType::Indirection` entries and skip merge resolution — blob pointer bytes are not user data. CompactionStream uses a pending buffer to preserve all consumed operands when hitting an Indirection base (no data loss); pending entries flow through the normal compaction pipeline (filter, GC, seqno zeroing). `resolve_merge_get` falls back to the latest operand instead of returning `Ok(None)`.

**Range reads**: `MvccStream` resolves merge operands during both forward and reverse iteration. Reverse iteration uses a reusable buffer (`key_entries_buf`) to avoid per-call allocation, and merges only when the newest entry is a `MergeOperand`.

**Implementor contract**: `MergeOperator::merge` must be deterministic and stable across multiple passes. `base_value` may already be the output of a previous merge (from compaction or earlier read resolution).

## Disk Format Compatibility

This PR participates in the fork's **disk format V4** boundary.

`V4` is required because the fork now persists semantics that a `V3` reader cannot interpret safely:

- **Range tombstones** add a new SST `range_tombstones` section / `BlockType::RangeTombstone`
- **Merge operators** add a persisted `ValueType::MergeOperand` in the on-disk value encoding

## API

```rust
pub trait MergeOperator: Send + Sync + RefUnwindSafe + 'static {
    fn merge(&self, key: &[u8], base_value: Option<&[u8]>, operands: &[&[u8]]) -> Result<UserValue>;
}

Config::new(path, seqno, visible_seqno)
    .with_merge_operator(Some(Arc::new(MyMergeOp)))
    .open()?;

tree.merge("counter", 1_i64.to_le_bytes(), seqno);
tree.get("counter", seqno)?; // returns merged result
```

## Known Limitations

- BlobTree read path does not yet resolve merge operands (compaction does); full BlobTree read-path support remains follow-up work
- Merge resolution logic is duplicated across `resolve_merge_get` (point reads) and `MvccStream` (range reads) — tracked in #46 for unification via bloom-filtered iterator pipeline

## Test Plan

- 40 integration tests: counter increments, base+merge, tombstone+merge, MVCC snapshots, flush/compaction, cross-layer merge, range scans (forward+reverse), prefix scan, multi_get, error propagation, BlobTree indirection fallback, major compaction, crash recovery, snapshot isolation, **range tombstone + merge** (RT kills base, RT kills all, RT in range scan, RT across flush, RT across multiple flushes, RT partial suppression, RT no-operator get/multi_get agreement)
- 16 `CompactionStream` unit tests: GC boundaries, partial merge type, complete vs partial, indirection base, error propagation, dropped callback, mixed watermark, filter type preservation
- 21 `MvccStream` unit tests: forward/reverse merge, indirection base, error propagation, weak tombstone, **RT suppression** (base, operand, head forward/reverse)
- 2 `Memtable` unit tests: `get_all_for_key` seqno=0 edge case, multi-version collection
- Full existing test suite (560+ tests) passes unchanged

Closes #15

## Summary by CodeRabbit

* **New Features**
  * Added merge operator support enabling custom merge logic for operands during reads and compaction operations.
  * Introduced `merge()` method to store merge operands that are combined with base values during access.
  * Configuration option to register custom merge operators per database instance for advanced use cases like counters and aggregations.

### CODE REVIEW

File: src/tree/mod.rs [hunk 1]
```diff
@@ -654,10 +674,26 @@ impl AbstractTree for Tree {
...
+                match entry {
+                    Some(entry) if entry.key.value_type == ValueType::MergeOperand => {
+                        if let Some(merge_op) = &self.config.merge_operator {
+                            Self::resolve_merge_get(
+                                &super_version,
+                                key,
+                                seqno,
+                                entry,
+                                merge_op.as_ref(),
+                            )
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/tree/mod.rs]: Fixed — resolve_merge_get redesigned to scan all sources from scratch, eliminating the duplicate operand bug and needless pass-by-value.

File: tests/merge_operator.rs [hunk 2]
```diff
@@ -0,0 +1,546 @@
+use lsm_tree::{AbstractTree, Config, Guard, MergeOperator, SequenceNumberCounter, UserValue};
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/merge_operator.rs]: Guard is required — `into_inner()` and `key()` are trait methods from `IterGuard` (re-exported as `Guard`). Removing the import causes compilation errors. Added clarifying comment.

**Comment by polaz** [comment]
@coderabbitai re-review

File: src/tree/mod.rs [hunk 3]
```diff
@@ -755,12 +820,175 @@ impl Tree {
...
+                        }
+                    } else if process_entry(&entry) {
+                        found_base = true;
+                    }
+
+                    if found_base {
+                        break;
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/tree/mod.rs]: The L0 overlap concern is tracked in #46 — code has a TODO(#46) comment documenting this limitation.

**Comment by polaz** [comment]
@coderabbitai re-review

**Comment by polaz** [comment]
@coderabbitai re-review

**Comment by polaz** [comment]
@coderabbitai re-review
